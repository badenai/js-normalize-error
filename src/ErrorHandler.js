/* global Response */
import isError from 'lodash/isError';

export default class ErrorHandler {
    constructor(isDebug = false, info) {
        this.isDebug = isDebug;
        this.info = info;
    }

    logError = (errorHandlerObject, info) => {
        let msg = 'logging error';
        if (info) msg += ` ${info}:\n`;
        else msg += `:\n`;
        Object.keys(errorHandlerObject).forEach(key => {
            if(key === 'headers') msg += `\t${key}: ${JSON.stringify(errorHandlerObject[key])}\n`;
            else msg += `\t${key}: ${errorHandlerObject[key]}\n`;
        });
        console.log(msg);
        return msg;
    };

    mapReadableError = errorHandlerObject =>
        errorHandlerObject.status
            ? {
                title: errorHandlerObject.title,
                status: errorHandlerObject.status,
                message: errorHandlerObject.message,
            }
            : {
                title: errorHandlerObject.title,
                message: errorHandlerObject.message,
            };

    getError = async error => {
        let errorHandlerObject = await this.handleError(error);
        if (!this.isDebug) {
            errorHandlerObject = this.mapReadableError(errorHandlerObject);
        }
        this.logError(errorHandlerObject, this.info);
        return errorHandlerObject;
    };

    mapJsError = error => ({
        title: error.name || typeof error,
        message: error.message || 'Unknown JS Error',
    });

    mapToObj = (map => {
        const obj = {};
        map.forEach((v, k) => { obj[k] = v });
        return obj;
    });

    getResponseObject = response => {
        let mappedResponse = {};
        Object.keys(response).forEach(key => {
            if (!key.startsWith('_')) mappedResponse[key] = response[key];
            if (key === 'headers') mappedResponse[key] = this.mapToObj(response[key]);
        })
        return mappedResponse;
    };

    mapJsonResponse = (response, json) => {
        const jsonResponse = this.getResponseObject(response);
        const mappedResponse = {
            ...jsonResponse,
            ...json,
            title: json.title || jsonResponse.statusText,
            status: jsonResponse.status,
            message: json.explanation,
        };
        return mappedResponse;
    };

    mapTextResponse = (response, text) => {
        const textResponse = this.getResponseObject(response);
        const mappedResponse = {
            ...textResponse,
            title: response.statusText || 'Response Error',
            status: textResponse.status,
            message: text,
        };
        return mappedResponse;
    };

    mapHtmlResponse = response => {
        const htmlResponse = this.getResponseObject(response);
        const mappedResponse = {
            ...htmlResponse,
            title: response.statusText || 'Response Error',
            status: htmlResponse.status,
            message: 'HTML can not be processed.',
        };
        return mappedResponse;
    };

    mapUnknownResponse = response => {
        const unknownResponse = this.getResponseObject(response);
        const mappedResponse = {
            ...unknownResponse,
            title: response.statusText || 'Unknown Response Error',
            status: unknownResponse.status,
            message: 'Something went wrong.',
        };
        return mappedResponse;
    };

    isJson = type => type.toLowerCase().includes('json');
    isText = type => type.toLowerCase().includes('plain');
    isHtml = type => type.toLowerCase().includes('html');

    handleError = async error => {
        if (isError(error)) {
            return this.mapJsError(error);
        } else if (error instanceof Response) {
            const response = error.clone();
            const contentType = response.headers.get('content-type');

            if (this.isJson(contentType)) {
                const json = await response.json();
                return this.mapJsonResponse(response, json);
            } else if (this.isText(contentType)) {
                const text = await response.text();
                return this.mapTextResponse(response, text);
            } else if (this.isHtml(contentType)) {
                return this.mapHtmlResponse(response);
            } else {
                return this.mapUnknownResponse(response);
            }
        } else {
            return {
                title: 'Unknown Error',
                message: 'Something went wrong.',
            };
        }
    };
}
