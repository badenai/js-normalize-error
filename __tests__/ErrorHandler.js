/* global Response, Headers*/
require('isomorphic-fetch');

import ErrorHandler from '../ErrorHandler';

const ResponseWrapper = (body, status, contentType) =>
    new Response(body, {
        status: status,
        headers: new Headers({ 'content-type': contentType }),
        url: 'https://somerandomurl.com',
    });

describe('ErrorHandler --- Method tests', () => {
    let errorHandler;
    beforeEach(() => {
        errorHandler = new ErrorHandler();
    });

    it('> map JSON Response', async () => {
        const response = ResponseWrapper(
            JSON.stringify({
                explanation: 'some error occurred',
                title: 'errorTitle',
            }),
            404,
            'application/json'
        );
        return expect(
            errorHandler.mapJsonResponse(response, await response.json())
        ).resolves.toMatchSnapshot();
    });
});

describe('ErrorHandler --- Get Readable Error Object', () => {
    let errorHandler;
    beforeEach(() => {
        errorHandler = new ErrorHandler(true);
    });

    it('> create instance', () => {
        expect(errorHandler).toBeDefined();
    });
    it('> json response', async () => {
        const response = ResponseWrapper(
            JSON.stringify({
                explanation: 'some error occurred',
                title: 'errorTitle',
            }),
            404,
            'application/json'
        );
        return expect(
            errorHandler.getError(response)
        ).resolves.toMatchSnapshot();
    });
    it('> text response', async () => {
        const response = ResponseWrapper(
            'some error occurred',
            404,
            'text/plain'
        );
        return expect(
            errorHandler.getError(response)
        ).resolves.toMatchSnapshot();
    });
    it('> html response', async () => {
        const response = ResponseWrapper(
            '<html><body></body></html>',
            404,
            'text/html'
        );
        return expect(
            errorHandler.getError(response)
        ).resolves.toMatchSnapshot();
    });
    it('> unknown response', async () => {
        const response = ResponseWrapper('', 404, 'test');
        return expect(
            errorHandler.getError(response)
        ).resolves.toMatchSnapshot();
    });
    it('> Type Error', async () => {
        const error = new TypeError('This is a type error');
        return expect(errorHandler.getError(error)).resolves.toMatchSnapshot();
    });
    it('> Error', async () => {
        const error = new Error('This is a error');
        return expect(errorHandler.getError(error)).resolves.toMatchSnapshot();
    });
    it('> unknown Error', async () => {
        const error = {};
        return expect(errorHandler.getError(error)).resolves.toMatchSnapshot();
    });
    it('> undefined', async () => {
        const error = undefined;
        return expect(errorHandler.getError(error)).resolves.toMatchSnapshot();
    });
});
