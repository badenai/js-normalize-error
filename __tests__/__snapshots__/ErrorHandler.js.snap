// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler --- Get Readable Error Object > Error 1`] = `
Object {
  "message": "This is a error",
  "title": "Error",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > Type Error 1`] = `
Object {
  "message": "This is a type error",
  "title": "TypeError",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > html response 1`] = `
Object {
  "_abort": false,
  "body": "<html><body></body></html>",
  "bodyUsed": false,
  "headers": Object {
    "content-type": Array [
      "text/html",
    ],
  },
  "message": "HTML can not be processed.",
  "ok": false,
  "size": 0,
  "status": 404,
  "statusText": "Not Found",
  "timeout": 0,
  "title": "Not Found",
  "url": "https://somerandomurl.com",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > json response 1`] = `
Object {
  "_abort": false,
  "_bytes": 58,
  "body": "{\\"explanation\\":\\"some error occurred\\",\\"title\\":\\"errorTitle\\"}",
  "bodyUsed": true,
  "explanation": "some error occurred",
  "headers": Object {
    "content-type": Array [
      "application/json",
    ],
  },
  "message": "some error occurred",
  "ok": false,
  "size": 0,
  "status": 404,
  "statusText": "Not Found",
  "timeout": 0,
  "title": "errorTitle",
  "url": "https://somerandomurl.com",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > text response 1`] = `
Object {
  "_abort": false,
  "_bytes": 19,
  "body": "some error occurred",
  "bodyUsed": true,
  "headers": Object {
    "content-type": Array [
      "text/plain",
    ],
  },
  "message": "some error occurred",
  "ok": false,
  "size": 0,
  "status": 404,
  "statusText": "Not Found",
  "timeout": 0,
  "title": "Not Found",
  "url": "https://somerandomurl.com",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > undefined 1`] = `
Object {
  "message": "Something went wrong.",
  "title": "Unknown Error",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > unknown Error 1`] = `
Object {
  "message": "Something went wrong.",
  "title": "Unknown Error",
}
`;

exports[`ErrorHandler --- Get Readable Error Object > unknown response 1`] = `
Object {
  "_abort": false,
  "body": "",
  "bodyUsed": false,
  "headers": Object {
    "content-type": Array [
      "test",
    ],
  },
  "message": "Something went wrong.",
  "ok": false,
  "size": 0,
  "status": 404,
  "statusText": "Not Found",
  "timeout": 0,
  "title": "Not Found",
  "url": "https://somerandomurl.com",
}
`;
